### PR TITLE
Add cdrom tray reboot/reset cases

### PR DESCRIPTION
**Pull request by chunfuwen**
Add cdrom tray reboot/reset cases
Bug 1824722 - cdrom tray state is not updated on VM reset

### CODE REVIEW

**Comment by chunfuwen** [comment]
Command 'virsh list' finished with 0 after 0.032575894s
cleaning libvirtd logs...
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virtual_disks.cdrom_device.coldplug.cdrom_reboot_reset_backend.positive_test

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py [hunk 1]
```diff
@@ -365,6 +366,60 @@ def create_block_lun_source_disk(params):
...
+    """
+    Check tray state can be updated after VM reboot or reset
+
+    :param vm: one object representing VM
+    :param params: wrapped parameters in dictionary format
+    :param test: test assert object
+    """
+    session = vm.wait_for_login()
+    eject_cdrom_device_cmd = "eject /dev/sr0"
+    status = session.cmd_status(eject_cdrom_device_cmd)
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:379]: Is it better to use cmd_status_output() to get the output for debugging?

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py [hunk 2]
```diff
@@ -365,6 +366,60 @@ def create_block_lun_source_disk(params):
...
+    if status:
+        test.fail("Failed to eject cdrom device in VM")
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:381]: Then include the output message here for debugging?

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py [hunk 3]
```diff
@@ -365,6 +366,60 @@ def create_block_lun_source_disk(params):
...
+        cdrom_xml = list(cdrom_devices.values())[0]
+        tray_state = cdrom_xml.find('target').get('tray')
+        LOG.debug("tray state value:\n %s", tray_state)
+
+        return tray_state
+
+    # tray state should be open
+    if _get_tray_state() != "open":
+        test.fail("Fail to eject guest cdrom, and expected tray state should be open, but actually: %s"
+                  % _get_tray_state())
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:398]: There is no need to execute this function twice. So I suggest maybe like below:
```
tray_state = _get_tray_state()
if tray_state != "open":
    test.fail("xxxxx" % tray_state)
```

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py [hunk 4]
```diff
@@ -365,6 +366,60 @@ def create_block_lun_source_disk(params):
...
+    # get qemu state
+    def _qemu_state():
+        qemu_output = virsh.qemu_monitor_command(name=vm.name, cmd="info block", options='--hmp',
+                                                 **{'debug': True, 'ignore_status': True})
+        LOG.debug("qemu output:\n%s", qemu_output.stdout_text)
+        return qemu_output.stdout_text
+
+    tray_original_state = params.get("tray_original_state")
+    if tray_original_state not in _qemu_state():
+        test.fail("Failed since tray state is open, but get state from qemu: %s" % _qemu_state())
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:409]: Same as above. No need to execute the function twice.

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py [hunk 5]
```diff
@@ -365,6 +366,60 @@ def create_block_lun_source_disk(params):
...
+    # reboot or reset VM
+    for virsh_function in [virsh.reboot, virsh.reset]:
+        virsh_function(vm.name, ignore_status=False, debug=True)
+        vm.wait_for_login().close()
+        # check tray state again, should be closed, and value is None
+        if _get_tray_state() is not None:
+            test.fail("Failed since expected tray state is closed, but get state is: %s" % _get_tray_state())
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:416]: Same as above

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py [hunk 6]
```diff
@@ -365,6 +366,60 @@ def create_block_lun_source_disk(params):
...
+        # check tray state from qemu
+        tray_reset_state = params.get("tray_reset_state")
+        if tray_reset_state not in _qemu_state():
+            test.fail("Failed since expect tray state is closed, but from qemu is: %s" % _qemu_state())
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:420]: same as above

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

**Comment by dzhengfy** [comment]
@qiankehan could you help review it?

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py [hunk 7]
```diff
@@ -365,6 +366,60 @@ def create_block_lun_source_disk(params):
...
+    def _qemu_state():
+        qemu_output = virsh.qemu_monitor_command(name=vm.name, cmd="info block", options='--hmp',
+                                                 **{'debug': True, 'ignore_status': True})
+        LOG.debug("qemu output:\n%s", qemu_output.stdout_text)
+        return qemu_output.stdout_text
+
+    tray_original_state = params.get("tray_original_state")
+    if tray_original_state not in _qemu_state():
+        test.fail("Failed since tray state is open, but get state from qemu: %s" % _qemu_state())
+    # reboot or reset VM
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:410]: 'or' should be 'and', right?

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py (hunk 4 above)
Comment (Yingshun) [reply libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:409]: And, is it better to make the word 'open' configurable? It's not easy to get the value of the 'tray_original_state' from L407~409.

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py [hunk 8]
```diff
@@ -365,6 +366,60 @@ def create_block_lun_source_disk(params):
...
+        """
+        cdrom_vmxml = vm_xml.VMXML.new_from_dumpxml(vm.name)
+        cdrom_devices = cdrom_vmxml.get_disk_all_by_expr('device==cdrom')
+        cdrom_xml = list(cdrom_devices.values())[0]
+        tray_state = cdrom_xml.find('target').get('tray')
+        LOG.debug("tray state value:\n %s", tray_state)
+
+        return tray_state
+
+    # tray state should be open
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:395]: Is this comment necessary?

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py [hunk 9]
```diff
@@ -365,6 +366,60 @@ def create_block_lun_source_disk(params):
...
+    if _get_tray_state() != "open":
+        test.fail("Fail to eject guest cdrom, and expected tray state should be open, but actually: %s"
+                  % _get_tray_state())
+
+    # get qemu state
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:400]: Can you delete this line and add doctstirng for _qemu_state()?

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py (hunk 1 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:379]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py (hunk 2 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:381]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py (hunk 6 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:420]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py (hunk 5 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:416]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py (hunk 4 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:409]: @dzhengfy  updated with avoiding execute function twice
@Yingshun  rename variable name: tray_original_state to tray_original_state_open

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py (hunk 7 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:410]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py (hunk 8 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_cdrom_device.py:395]: removed

**Comment by chunfuwen** [comment]
@dzhengfy 
@Yingshun 
comment are addressed, please double check

**Review by Yingshun — APPROVED** [review]

**Comment by qiankehan** [comment]
For this case, expect the **tray-change** event emitted when reboot/reset after the tray status changed.  Could you add the check for this event?

**Comment by chunfuwen** [comment]
@qiankehan If we look at original issue in https://bugzilla.redhat.com/show_bug.cgi?id=1824722, what we need validate is tray state can be reflected correctly. Currently cases checking already includes vm xml, virsh qemu-monitor-command ,and VM internal checking. Therefore it is enough to ensure cdrom tray state validation.  Virsh events is here only auxiliary tool to help validate cdrom tray state. Considering previous three layers validation. Look like virsh  events is not necessary here.

**Comment by qiankehan** [comment]
> @qiankehan If we look at original issue in https://bugzilla.redhat.com/show_bug.cgi?id=1824722, what we need validate is tray state can be reflected correctly. Currently cases checking already includes vm xml, virsh qemu-monitor-command ,and VM internal cdrom device checking. Therefore here they are enough to ensure cdrom tray state validation. Virsh events is here only auxiliary tool to help validate cdrom tray state. Considering previous three layers validation. Look like virsh events is not necessary here.

In fact, [the original issue](https://bugzilla.redhat.com/show_bug.cgi?id=1824722) also mentioned that the tray-change events is not emitted:
> state of cdrom tray is not updated (changed to 'closed') when resetting a VM, but the internal tray state in **qemu is actually changed to 'closed' without emission of the tray-changed event**.

And I verified that at https://bugzilla.redhat.com/show_bug.cgi?id=1824722#c20 https://bugzilla.redhat.com/show_bug.cgi?id=1824722#c22 and https://bugzilla.redhat.com/show_bug.cgi?id=1824722#c26. The developer fixed this issue at https://bugzilla.redhat.com/show_bug.cgi?id=1824722#c21

That's why I think we should check the tray-change event here.

**Comment by chunfuwen** [comment]
> > @qiankehan If we look at original issue in https://bugzilla.redhat.com/show_bug.cgi?id=1824722, what we need validate is tray state can be reflected correctly. Currently cases checking already includes vm xml, virsh qemu-monitor-command ,and VM internal cdrom device checking. Therefore here they are enough to ensure cdrom tray state validation. Virsh events is here only auxiliary tool to help validate cdrom tray state. Considering previous three layers validation. Look like virsh events is not necessary here.
> 
> In fact, [the original issue](https://bugzilla.redhat.com/show_bug.cgi?id=1824722) also mentioned that the tray-change events is not emitted:
> 
> > state of cdrom tray is not updated (changed to 'closed') when resetting a VM, but the internal tray state in **qemu is actually changed to 'closed' without emission of the tray-changed event**.
> 
> And I verified that at https://bugzilla.redhat.com/show_bug.cgi?id=1824722#c20 https://bugzilla.redhat.com/show_bug.cgi?id=1824722#c22 and https://bugzilla.redhat.com/show_bug.cgi?id=1824722#c26. The developer fixed this issue at https://bugzilla.redhat.com/show_bug.cgi?id=1824722#c21
> 
> That's why I think we should check the tray-change event here.

@qiankehan the tray-change event not emission is the root clause where libvirt VM xml is not consistent with qemu side(this is bug' addressing issue). So from E2E point view,  libvirt xml state and qemu side are End to End result in this specific case, not the event emission. 
But, anyway, I can add this additional checkpoint

**Comment by chunfuwen** [comment]
@qiankehan additional event type is added
New test result is uploaded:

Command 'virsh list' finished with 0 after 0.032728677s
cleaning libvirtd logs...
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virtual_disks.cdrom_device.coldplug.cdrom_reboot_reset_backend.positive_test

**Review by dzhengfy — APPROVED** [review]

**Pull request merged**